google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -128,12 +100,22 @@ public String toString() {
     private final int outstandingBytes;
     private final long receivedTimeMillis;
     private final Instant totalExpiration;
+    private final SettableApiFuture<AckResponse> messageFuture;
 
-    private AckHandler(String ackId, int outstandingBytes, Instant totalExpiration) {
+    private AckHandler(
+        String ackId,
+        int outstandingBytes,
+        Instant totalExpiration,
+        SettableApiFuture<AckResponse> messageFuture) {
       this.ackId = ackId;
       this.outstandingBytes = outstandingBytes;
       this.receivedTimeMillis = clock.millisTime();
       this.totalExpiration = totalExpiration;
+      this.messageFuture = messageFuture;
```

Comment:
We seem to be creating AckIdMessageFutures all over the place, using the same ackId and same messageFuture stored in the AckHandler. Should we intead just store an AckIdMessageFuture in the AckHandler, which contains the ackId and messageFuture?

Reply:
We should be creating one `AckIdMessageFuture` here: https://github.com/googleapis/java-pubsub/pull/1022/files#diff-34fea2118b399ecaf7c286bc8f4920c30bb6a6ab0f732a922e1829d65c11643bR302 (line 302 in `MessageDispatcher.java` - this file, if that link doesn't work), and then passing it around.

I could move the failure logic to the AckHandler, but thought it was easier to understand/read to have the failure logic be in the StreamingSubscriberConnection versus throwing an exception to be caught by the individual AckHandlers. I am not opposed to moving it though if you think that makes more sense.

Reply:
Good call, will do.